File: IN/WEEK01/BOJ1629곱셈_TIN.py
import sys

input = sys.stdin.readline

num, exp, div = map(int, input().split())

def pow(num, exp, div):
    if exp == 0:
        return 1
    tmp = pow(num, exp//2, div)
    tmp = (tmp * tmp) % div
    if exp % 2 == 0:
        return tmp
    return (num * tmp) % div 
    

print(pow(num, exp, div))        

File: IN/WEEK10/BOJ1629곱셈_TIN.cpp
#include <iostream> 
#include <vector>
#include <algorithm>
typedef long long ll;
using namespace std;

ll pow(ll num, ll exp, ll div);

int main(void){
    ll num, exp, div;
    scanf("%lu %lu %lu", &num, &exp, &div);
    printf("%lu", pow(num, exp, div));

    return 0;
}

ll pow(ll num, ll exp, ll div){
    if(exp == 0) return 1;
    ll tmp = pow(num, exp / 2, div);
    tmp = (tmp * tmp) % div;
    if(exp % 2 == 0) return tmp;
    return (num * tmp) % div;
}